Mathlib/Order/KrullDimension.lean

```diff
@@ -30,40 +40,104 @@ in this file would generalize as well. But we don't think it would be useful, so
 Krull dimension of a preorder.
 -/
 
-section definitions
+namespace Order
 
-variable (α : Type*) [Preorder α]
+section definitions
 
 /--
 The **Krull dimension** of a preorder `α` is the supremum of the rightmost index of all relation
-series of `α` order by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
+series of `α` ordered by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
 dimension is defined to be negative infinity; if the length of all series `a₀ < a₁ < ... < aₙ` is
 unbounded, its Krull dimension is defined to be positive infinity.
 -/
-noncomputable def krullDim : WithBot (WithTop ℕ) :=
+noncomputable def krullDim (α : Type*) [Preorder α] : WithBot ℕ∞ :=
   ⨆ (p : LTSeries α), p.length
 
 /--
-Height of an element `a` of a preordered set `α` is the Krull dimension of the subset `(-∞, a]`
+The **height** of an element `a` in a preorder `α` is the supremum of the rightmost index of all
+relation series of `α` ordered by `<` and ending below or at `a`.
 -/
-noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a)
+noncomputable def height {α : Type*} [Preorder α] (a : α) : ℕ∞ :=
+  ⨆ (p : LTSeries α) (_ : p.last ≤ a), p.length
 
 /--
-Coheight of an element `a` of a pre-ordered set `α` is the Krull dimension of the subset `[a, +∞)`
+The **coheight** of an element `a` in a preorder `α` is the supremum of the rightmost index of all
+relation series of `α` ordered by `<` and beginning with `a`.
+
+The definition of `coheight` is via the `height` in the dual order, in order to easily transfer
+theorems between `height` and `coheight`.
 -/
-noncomputable def coheight (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Ici a)
+noncomputable def coheight {α : Type*} [Preorder α] (a : α) : ℕ∞ := height (α := αᵒᵈ) a
 
 end definitions
 
+section height
+
+variable {α β : Type*}
+
+variable [Preorder α] [Preorder β]
+
+/-!
+## Height
+-/
+
+lemma height_le_iff (x : α) (n : ℕ∞) :
+    height x ≤ n ↔ ∀ (p : LTSeries α), p.last ≤ x → p.length ≤ n := by
+  simp [height, iSup_le_iff]
+
+lemma height_le (x : α) (n : ℕ∞) :
+    (∀ (p : LTSeries α), p.last ≤ x → p.length ≤ n) → height x ≤ n :=
+  (height_le_iff x n).mpr
+
+lemma length_le_height (x : α) (p : LTSeries α) (hlast : p.last ≤ x) :
+    p.length ≤ height x := by
+  by_cases hlen0 : p.length ≠ 0
+  · let p' := p.eraseLast.snoc x (by
+      apply lt_of_lt_of_le
+      · apply p.step ⟨p.length - 1, by omega⟩
+      · convert hlast
+        simp only [Fin.succ_mk, Nat.succ_eq_add_one, RelSeries.last, Fin.last]
+        congr; omega)
+    suffices p'.length ≤ height x by
+      simp [p'] at this
+      convert this
+      norm_cast
+      omega
+    refine le_iSup₂_of_le p' ?_ le_rfl
+    simp [p']
+  · simp_all
+
+lemma length_le_height_last (p : LTSeries α) : p.length ≤ height p.last :=
```

Comment:
Even here I would make `p` implicit. It can be figured out from either side of the inequality
```suggestion
lemma length_le_height_last : p.length ≤ height p.last :=
```

Reply:
Looks like I am using this fairly often in a forward-reasoning style, e.g. with `have`. But I can use `(p := …)` then, so all good.